src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs

```diff
@@ -747,8 +747,22 @@ public BoundExpression Sequence(params BoundExpression[] parts)
         public BoundExpression Sequence(ImmutableArray<LocalSymbol> locals, params BoundExpression[] parts)
         {
             var builder = ArrayBuilder<BoundExpression>.GetInstance();
-            for (int i = 0; i < parts.Length - 1; i++) builder.Add(parts[i]);
+            for (int i = 0; i < parts.Length - 1; i++)
+            {
+                var part = parts[i];
+                if (LocalRewriter.ReadIsSideeffecting(part))
```

Comment:
It would be fine to have this optimization in the LocalRewriter. This class is a factory, it should simply create what it is asked for, or, at least, the name of the API should be adjusted. Perhaps a new API.

---
In reply to: [77089404](https://github.com/dotnet/roslyn/pull/13501#discussion_r77089404) [](ancestors = 77089404)